### ISSUE TITLE
Implement non-causal reference attention (stable softmax)

### ISSUE THREAD

**Issue by sreehariam** [sreehariam/fa-research-cpp#1]
Add a CPU reference implementation of fa::attention_forward (non-causal, no padding mask). Compute O(N²) logits with numerically stable softmax (max-trick) and return (B,H,N,D).

```
// include/fa/attention.hpp
namespace fa {
Tensor attention_forward(const Tensor& Q,
                         const Tensor& K,
                         const Tensor& V,
                         const Tensor* mask,          // ignored in this issue
                         const AttentionOpts& opts);  // opts.causal must be false
}
```

**Inputs & shapes**
Q,K,V: (B,H,N,D); all 4D; same B,H,N,D. mask=nullptr. opts.causal=false.

**Behavior**
- logits: l_j = dot(Q[b,h,i,:], K[b,h,j,:])
- softmax: p_j = exp(l_j - max(l))/sum_k exp(l_k - max(l))
- output: O[b,h,i,:] = Σ_j p_j * V[b,h,j,:]

**Acceptance**

- Stable softmax; finite outputs; correct shape.
- Throws std::invalid_argument on shape mismatch.
- Matches a small deterministic case (uniform softmax when Q=K=0).
- Row probs sum ≈ 1 (checked indirectly).

**Out of scope**
Padding mask & causal attention; CUDA.
